### PR TITLE
Add telemetry/event support

### LINKED ISSUE #430: Add telemetry/event support
collect at least the startup event with 
- Server Version
- JVM Version
- memory

### PR DESCRIPTION
**Pull request by angelozerr**
This PR fix #430

The OSUtils was renamed to Platform which is more generic (eclipse like
name) which contains information about OS, JVM (name, version, memory).

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+package org.eclipse.lsp4xml;
+
+/**
+ * PLatform information about OS and JVM.
+ */
+public class Platform {
+
+	private static final OS os = new OS();
+
+	private static final JVM jvm = new JVM();
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java]: no need to keep it static, as it should be used once on startup

**Comment by angelozerr** [comment]
@fbricon here my vscode trace:

```json
[Trace - 15:05:19] Received notification 'telemetry/event'.
Params: {
    "version": "0.8.0-20190701-1255",
    "os": {
        "name": "Windows 10",
        "version": "10.0",
        "arch": "amd64"
    },
    "jvm": {
        "name": "OpenJDK 64-Bit Server VM",
        "version": "1.8.0_212-3-redhat",
        "memory": {
            "free": 520615944,
            "total": 534773760,
            "max": 1073741824
        }
    }
}

```

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java]: I agree for JVM, but for os it used to initialize isWindows static variable

Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java]: yes, that's why I only commented for JVM

**Comment by angelozerr** [comment]
@fbricon I have add telemetry to track explicit action from user, that's to say:

 * rename
 * formatting
 * rangeFormatting

In vscode trace, you should see :

```
[Trace - 09:44:17] Received notification 'telemetry/event'.
Params: "textDocument/formatting"
```

if you format a XML file.

**Comment by angelozerr** [comment]
> Looks OK, but please add some unit tests before we merge it.

It will hard to do that since it depends on OS where JUnit are executed (I think about JVM and OS information).

Or perhaps you wish I write test just with rename, formatting, rangeFormatting, and onInitialized but by avoiding value check of OS, JVM memories?

An another question, I tell me if we should give the capability to send telemetry or not. Here some ideas:

 * add settings to disable telemetry?
 * add client capability 'telemetry' to enable telemetry. For instance I think LSP4E will never used those telemetry events.

**Comment by fbricon** [comment]
You can check telemetry data contain the same values as what's in Platform, and then check memory values are > 0.

Yes an `xml.telemetry.enabled` preference would be useful.

**Comment by angelozerr** [comment]
> Looks OK, but please add some unit tests before we merge it.

Done with https://github.com/angelozerr/lsp4xml/pull/493/files#diff-06ace1651d73790f809dd84073bc3a86R45

**Comment by angelozerr** [comment]
> You can check telemetry data contain the same values as what's in Platform, and then check memory values are > 0.

Ok test was updated.

> Yes an xml.telemetry.enabled preference would be useful.

Done in PR https://github.com/redhat-developer/vscode-xml/pull/165

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLTelemetrySettings.java [hunk 2]
```diff
@@ -8,15 +8,21 @@
...
-public class OSUtils {
-	
-	public static final boolean isWindows = System.getProperty("os.name").toLowerCase().indexOf("win") >= 0;
-	public static String SLASH = isWindows ? "\\" : "/";
+public class XMLTelemetrySettings {
 
-	
-}
\ No newline at end of file
+	private boolean enabled = true;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLTelemetrySettings.java]: should be off by default, clients should explicitly enable it

Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLTelemetrySettings.java]: Good catch! It was a compy from XMLSymbolSettings. Copy/paste generate so many errors :)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	 * Returns the system property from the given key and null otherwise.
+	 * 
+	 * @param key the property system key
+	 * @return the system property from the given key and null otherwise.
+	 */
+	private static String getSystemProperty(String key) {
+		try {
+			return System.getProperty(key);
+		} catch (SecurityException e) {
+			return null;
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java]: @fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	 */
+	public static OS getOS() {
+		return os;
+	}
+
+	/**
+	 * Returns the system property from the given key and null otherwise.
+	 * 
+	 * @param key the property system key
+	 * @return the system property from the given key and null otherwise.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/Platform.java]: @fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/telemetry/TelemetryManager.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+	/**
+	 * The telemetry notification is sent from the server to the client to ask the
+	 * client to log a telemetry event.
+	 */
+	private void telemetryEvent(Object object) {
+		languageClient.telemetryEvent(object);
+	}
+
+}
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/telemetry/TelemetryManager.java]: @fbricon updated

**Comment by angelozerr** [comment]
Replaced with https://github.com/eclipse/lemminx/pull/989

**Pull request closed without merging**